### PR TITLE
Support Inputs and Labels in the dict format

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- flatten dict structure before saving them as tensors.
- Using the same logic used by tf to handle dict structures.

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/tensorflow2/test_support_dicts.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+from smdebug.core.collection import CollectionKeys
+from smdebug.trials import create_trial
+
+
+def get_data():
+    images = np.zeros((64, 224))
+    labels = np.zeros((64, 5))
+    inputs = {"Image_input": images}
+    outputs = {"output-softmax": labels}
+    return inputs, outputs
```
Comment (vandanavk) [inline tests/tensorflow2/test_support_dicts.py:16]: can we have some test cases for valid and invalid dicts?
empty dicts,
either inputs/output empty (is this possible?)
are the values always valid? - is there a chance they cant be converted to numpy?

is this applicable only to image inputs?

File: tests/tensorflow2/test_support_dicts.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    return model
+
+
+def test_support_dicts(out_dir):
+    model = create_model()
+    optimizer = tf.keras.optimizers.Adadelta(lr=1.0, rho=0.95, epsilon=None, decay=0.0)
+    model.compile(loss="categorical_crossentropy", optimizer=optimizer)
+    inputs, labels = get_data()
+    smdebug_hook = create_hook(out_dir)
+    model.fit(inputs, labels, batch_size=16, epochs=10, callbacks=[smdebug_hook])
```
Comment (vandanavk) [inline tests/tensorflow2/test_support_dicts.py:38]: save the model at the end. just to check it works fine

File: tests/tensorflow2/test_support_dicts.py (hunk 1 above)
Comment (vandanavk) [reply tests/tensorflow2/test_support_dicts.py:16]: are dicts the only other valid format? can inputs be nested lists or something similar?

File: smdebug/tensorflow/keras.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 import tensorflow.compat.v1 as tf
 from tensorflow.python.distribute import values
 from tensorflow.python.framework.indexed_slices import IndexedSlices
+from tensorflow.python.util import nest
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:8]: is this valid in both TF1 and TF2?

File: smdebug/tensorflow/keras.py [hunk 4]
```diff
@@ -497,6 +498,10 @@ def save_smdebug_logs(self, logs):
                         else set()
                     )
                 for t_name, t_value in tensors_to_save:
+                    if isinstance(t_value, dict):
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:501]: can it get to this point in non-eager mode?

File: tests/tensorflow2/test_support_dicts.py (hunk 1 above)
Comment (vandanavk) [reply tests/tensorflow2/test_support_dicts.py:16]: is this used only with keras fit or can it be used with gradtape too?

File: smdebug/tensorflow/keras.py (hunk 3 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:8]: This import is valid in both TF1 and TF2

File: smdebug/tensorflow/keras.py (hunk 4 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:501]: Both eager and non-eager modes execute this code-block.

File: tests/tensorflow2/test_support_dicts.py (hunk 1 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_support_dicts.py:16]: The flatten call made by smdebug is only after tf runs the same function on the feed_dict.

Both fit and gradtape execute the same call stack

File: tests/tensorflow2/test_support_dicts.py (hunk 2 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_support_dicts.py:38]: sure.

File: tests/tensorflow2/test_support_dicts.py (hunk 1 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_support_dicts.py:16]: > can we have some test cases for valid and invalid dicts?
> empty dicts,
> either inputs/output empty (is this possible?)
> are the values always valid? - is there a chance they cant be converted to numpy?
> 
> is this applicable only to image inputs?

model.fit breaks if the customer passes an empty or invalid feed_dict.

**Review by vandanavk — APPROVED** [review]
